### DISCUSSION TITLE
Mise support in renovate

### DISCUSSION THREAD

**Discussion by jvizier** [jdx/mise#1940] [Ideas]
Hi and thanks for mise :)

Actually asdf is already supported in [Renovate](https://docs.renovatebot.com/modules/manager/asdf/)

- It is possible to have the same for mise ?

- There is already an open discussion : https://github.com/renovatebot/renovate/discussions/26596

- There is a questionnaire to fo fill : https://github.com/renovatebot/renovate/blob/main/docs/development/new-package-manager-template.md

Any help would be appreciated ( i will try asap )

Best regards.

**Comment by jdx** [comment]
here, I'll fill this out for you:

# New package manager questionnaire

Did you read our documentation on adding a package manager?

- [ ] I've read the [adding a package manager](adding-a-package-manager.md) documentation.

## Basics

### What's the name of the package manager?

"mise-en-place" or simply "mise"

### What language(s) does this package manager support?

Virtually all of them.

### How popular is this package manager?

It's ~200th place in Homebrew across all formulas.

### Does this language have other (competing?) package managers?

- [x] Yes: asdf, vfox, proto, aqua
- [ ] No.

### What are the big selling points for this package manager?

Performance, developer experience. Mise's goal is to become the "front-end to your dev env" with the ultimate goal of being the standard choice for developers at organizations and in open source projects to adopt for running their codebases locally.

Explain how this package manager is different from existing ones.

Mise focuses on DX first. Other projects try to optimize for things like stability, cross-platform support, or hardened security models—while these are also things mise strives to achieve they come second after DX at least right now. Compared to asdf, mise is a more ambitious project with a broader feature set targeting things like dev environment management and task execution.

## Detecting package files

### What kind of package files, and names, does this package manager use?

* asdf's `.tool-versions` files
* mise's `.mise.toml` files
* "legacy version files" https://mise.jdx.dev/configuration.html#legacy-version-files

### Which [`fileMatch`](../usage/configuration-options.md#filematch) pattern(s) should Renovate use?

Probably just `.mise.toml` and `.config/mise/config.toml` to start, there are others but not all would make sense in the context of renovate (like `.mise.local.toml` and `.mise.staging.toml`).

### Do many users need to extend the [`fileMatch`](../usage/configuration-options.md#filematch) pattern for custom file names?

- [x] Possibly, users can use custom filenames, or "environment-specific" config files like `.mise.production.toml`.
- [ ] No.

### Is the [`fileMatch`](../usage/configuration-options.md#filematch) pattern going to get many "false hits" for files that have nothing to do with package management?

Not for `.mise.toml`. I doubt anyone would use that name for anything else and mise would throw errors if it found one.

## Parsing and Extraction

### Can package files have "local" links to each other that need to be resolved?

Some translation needs to be done here since renovate is for package management but mise isn't a package manager, it's a tool manager. For that there is some capability for using `path:~/some/path` to use tools in a local directory. This isn't a super popular feature, and it's usually used for custom compiled tools—not something where a tool is inside of a project. That said, I'm sure at some point people would try to accomplish something like this.

That said, I don't think renovate needs to do anything special for that aside from just knowing it's a valid configuration. It probably makes sense just to not touch it.

### Package file parsing method

The package files should be:

- [ ] Parsed together (in serial).
- [ ] Parsed independently.
I don't really understand what this is asking. I think it's whether each config file can be parsed independently to which the answer is "yes". That said, it is possible for parsing of files to change based on when the parsing happens... but this is a pretty complex behavior and I don't see it impacting renovate in any significant way. This would be around something like having "poetry" in a local config file that uses "python" from a global one. This couldn't be reversed since poetry depends on python being installed.

### Which format/syntax does the package file use?

- [ ] JSON
- [x] TOML
- [ ] YAML
- [ ] Custom (explain below)

### How should we parse the package files?

- [x] Off the shelf parser.
- [ ] Using regex.
- [ ] Custom-parsed line by line.
- [ ] Other.

### Does the package file have different "types" of dependencies?

- [ ] Yes, production and development dependencies.
- [x] No, all dependencies are treated the same. — though separate `.mise.production.toml` files are supported. That's just an arbitrary name that is read from the `MISE_ENV` env var.

### List all the sources/syntaxes of dependencies that can be extracted

### Describe which types of dependencies above are supported and which will be implemented in future

## Versioning

### What versioning scheme does the package file(s) use?

One that is difficult to explain. It's sort of like a "prefix" match where "20" matches to things like "20.1.0" or "20.0.1" but not "21.0". The logic is quite complex in order to work reasonably well across every language. As of this writing the logic _is_ consistent across languages but I believe that will not always be the case. Some languages likely will end up using completely different logic.

### Does this versioning scheme support range constraints, like `^1.0.0` or `1.x`?

- [x] Supports range constraints (for example: `^1.0.0` or `1.x`), provide details.
- [ ] No.

## Lookup

### Is a new datasource required?

- [x] Yes, provide details.
- [ ] No.

It's different for each language. There is a service that mise uses internally to "cache" versions of popular languages. I'm not sure it's a good idea to use it for renovate since I am not ready to guarantee that it will always be around or have the same behavior.

### Will users want (or need to) set a custom host or custom registry for Renovate's lookup?

- [ ] Yes, provide details.
- [x] No.

This is also different for each language. Some languages provide a mechanism for this, some don't. There isn't consistency in how that configuration is done either.

Where can Renovate find the custom host/registry?

- [x] No custom host or registry is needed.
- [ ] In the package file(s), provide details.
- [ ] In some other file inside the repository, provide details.
- [ ] User needs to configure Renovate where to find the information, provide details.

### Are there any constraints in the package files that Renovate should use in the lookup procedure?

- [ ] Yes, there are constraints on the parent language (for example: supports only Python `v3.x`), provide details.
- [ ] Yes, there are constraints on the parent platform (for example: only supports Linux, Windows, etc.), provide details.
- [ ] Yes, some other kind of constraint, provide details.
- [x] No constraints.

### Will users need the ability to configure language or other constraints using Renovate config?

- [ ] Yes, provide details.
- [x] No.

## Artifacts

### Does the package manager use a lock file or checksum file?

- [ ] Yes, uses lock file.
- [ ] Yes, uses checksum file.
- [ ] Yes, uses lock file _and_ checksum file.
- [x] No lock file or checksum.

### Is the locksum or checksum mandatory?

- [ ] Yes, locksum is mandatory.
- [ ] Yes, checksum is mandatory.
- [ ] Yes, lock file _and_ checksum are mandatory.
- [ ] No mandatory locksum or checksum.
- [x] Package manager does not use locksums or checksums.

### If lockfiles or checksums are used: what tool and exact commands should Renovate use to update one (or more) package versions in a dependency file?

### Package manager cache

#### Does the package manager use a cache?

- [x] Yes, provide details.
- [ ] No.

Mise has an internal cache. It's not intended to be used by anything else. Perhaps this could be revisited if we had a use-case for making the cache distributable or something though.

#### If the package manager uses a cache, how can Renovate control the cache?

- [ ] Package manager does not use a cache.
- [ ] Controlled via command line interface, provide details.
- [ ] Controlled via environment variables, provide details.

#### Should Renovate keep a cache?

- [ ] Yes, ignore/disable the cache.
- [ ] No.

Not sure what a renovate cache is frankly.

### Generating a lockfile from scratch

Renovate can perform "lock file maintenance" by getting the package manager to generate a lockfile from scratch.
Can the package manager generate a lockfile from scratch?

- [ ] Yes, explain which command Renovate should use to generate the lockfile.
- [ ] No, the package manager does _not_ generate a lockfile from scratch.
- [ ] No, the package manager does not use lockfiles.

Mise does not _currently_ use lockfiles but this is an idea that's come up several times to improve reproducibility. I think mise will have it at some point but there aren't near-term plans for this.

## Other

### What else should we know about this package manager?

It may be worth exploring how renovate could integrate with tasks—either by having a `mise run renovate` task or a hook for renovate to execute tasks. Happy to brainstorm on this as tasks are new and can very much still be improved.

**Comment by jvizier** [comment]
Thanks @jdx for the work :) i will update the thread on renovate.

update : https://github.com/renovatebot/renovate/discussions/26596

**Reply by reitzig** [reply]
Apparently, support was added mid-2024 with `37.426.0` (cf. [renovatebot/renovate#28918.2214319228](https://github.com/renovatebot/renovate/issues/28918#issuecomment-2214319228)); documentation at [Modules > Managers > mise](https://docs.renovatebot.com/modules/manager/mise/).

I was thinking that, maybe, providing a shared preset the way aqua does (cf. [aquaproj/aqua-renovate-config](https://github.com/aquaproj/aqua-renovate-config)) would be more sustainable due to the many datasources that are relevant for mise, and the high speed of development that may be hard to match for the Renovate maintainers. 
For instance, we already see that the current default backend (aqua, if I'm not mistaken) doesn't seem to be supported.

I'd be interested in experiences before migrating some aqua configs I have over to mise, as getting updates via Renovate is very important to me in dev setups. :)

**Reply by reitzig** [reply]
That said, not being limited to `regex`-managers is certainly a plus!

**Reply by jdx** [reply]
I'm definitely not going to take that on

**Reply by reitzig** [reply]
Completely fair. 

That said, I strongly suspect your input will be necessary: between multiple backends (all with their own Renovate datasources), aliases (that can _hide_ the backend used!), and soon lock files, the logic to resolve which datasource to use for which tool entry may become hard to formulate. 

_Maybe_ Renovate should just delegate to something like `mise upgrade --bump --no-install` (if that were a thing; am I missing something?); not sure how that would integrade with the concept of datasources in Renovate, would mise be its own? 🤔 #thinking-aloud

**Reply by zeitlinger** [reply]
Related suggestion for renovate: https://github.com/jdx/mise/discussions/4682

**Comment by risu729** [comment]
I opened a PR to add backend support last December, but it hasn’t been merged yet:
https://github.com/renovatebot/renovate/pull/33113.

In the meantime, I’ve been using a regex manager like this:
[risu729/renovate-config](https://github.com/risu729/renovate-config/blob/main/src/default.json5)
but it has limitations and doesn’t support many use cases.

**Comment by jylenhof** [comment]
Maybe you can use a derivative of this github workflow I wrote which use mise itself to generate a PR by tool :
https://github.com/jylenhof/mise-update-by-tool

**Reply by zeitlinger** [reply]
great :smile: 

1. can this be used directly?
2. can you add a README.md?

**Reply by jylenhof** [reply]
Yes. Feel free to ameliorate it and/or push some PRs, this is Open Source ! 

Principal default is that the PR will not trigger your check workflow because of rights, you probably need an OIDC Pat github token (we made a variant of this at work I cannot publish)

Other ones :
-  mise does'nt handle idiomatic file for update, so you should add a mise.local.toml file that desactivate idiomatic files in a step
-  if everything is up-to-date, workflow should not fail (if is missing on the second part of the workflow)

**Comment by zeitlinger** [comment]
Mise is already supported in some way - so this is about extending the support I guess.

https://github.com/renovatebot/renovate/pull/29950

Here's my PR that bumps: https://github.com/grafana/docker-otel-lgtm/pull/372

**Comment by reitzig** [comment]
Renovate 40.30.0 contained a significant update for mise support: https://github.com/renovatebot/renovate/pull/33113#issuecomment-2907645456

Thanks, @risu729 ! ♥️

**Comment by zeitlinger** [comment]
would it be time for lockfile support in renovate now?

**Reply by jdx** [reply]
I wouldn't until it comes out of experimental

**Reply by risu729** [reply]
Just as a note, generally Renovate runs the `install` command to update the lockfiles.
So I think we should have `--write-lock` or something like that to only update the lockfile, but not installing the tools because installing all tools will be super slow and I don't think Renovate will accept such features.

**Reply by zeitlinger** [reply]
`--write-lock` is a bit confusing - maybe `--write-lockfile-only`

**Comment by zeitlinger** [comment]
I just noticed that renovate doesn't update `mise.<env>.toml` - I guess that would be easy to add. 

@jdx does this make sense?

**Reply by risu729** [reply]
I intentionally excluded them in https://github.com/renovatebot/renovate/pull/30283, but I agree that we can update those configs with Renovate.
